src/main/java/com/nvs/store/service/CartService.java

```diff
@@ -0,0 +1,202 @@
+        User user = userRepository
+                .findByEmail(userEmail)
+                .orElseThrow(() -> new RuntimeException(
+                        String.format("User with email [%s] not found", userEmail))
+                );
+        return cartRepository
+                .findByUserId(user.getId())
+                .stream()
+                .map(CartUtils.convertToCartDto)
+                .toList();
+    }
+
+    public CartDTO addProductToCart(CartItemRequest cartItemRequest) {
+        String userEmail = getUserEmailFromContext();
+        User user = userRepository.findByEmail(userEmail)
+                .orElseThrow(() -> new RuntimeException(
+                        String.format("User with username %s not found", userEmail))
+                );
+
+        Product product = productRepository
+                .findById(cartItemRequest
+                        .getProductId())
+                .orElseThrow(() ->
+                        new RuntimeException(
+                                String.format("Product with id [%d] does not exist", cartItemRequest.getProductId()))
+                );
+        Cart cart = cartRepository
+                .findByUserIdAndActualTrue(user.getId());
+
+        List<CartItem> cartItems = cartItemRepository.findCartItemByCart(cart);
+        for (CartItem cartItem : cartItems) {
+            if (cartItem.getProduct().getId().equals(product.getId())) {
+                int newQuantity = cartItem.getQuantity() + cartItemRequest.getQuantity();
+                validateAvailableProductQuantities(product, newQuantity);
+                cartItem.setQuantity(newQuantity);
+                cartItemRepository.save(cartItem);
+                return CartUtils.convertToCartDto.apply(cartRepository.findByUserIdAndActualTrue(user.getId()));
+            }
+        }
+
+        validateAvailableProductQuantities(product, cartItemRequest.getQuantity());
+        CartItem cartItem = CartItem
+                .builder()
+                .quantity(cartItemRequest.getQuantity())
+                .product(product)
+                .cart(cart)
+                .build();
+
+        cartItemRepository.save(cartItem);
+
+        return CartUtils.convertToCartDto
+                .apply(cartRepository
+                        .findByUserIdAndActualTrue(user.getId()));
+    }
+
+    public CartDTO orderCart() {
+        String userEmail = getUserEmailFromContext();
+        User user = userRepository.findByEmail(userEmail).orElseThrow(
+                () -> new RuntimeException(
+                        String.format("User with username %s not found", userEmail))
+        );
+
+        Cart cart = cartRepository
+                .findByUserIdAndActualTrue(user.getId());
+
+        List<CartItem> orderedCartItems = cartItemRepository.findCartItemByCart(cart);
+
+        for (CartItem cartItem : orderedCartItems) {
+            Product product = productRepository.findById(cartItem.getProduct().getId())
```

Comment:
The operation is redundant because the method call productRepository.findById(cartItem.getProduct().getId()) to return Optional#empty will not occur. The database schema should to ensure that a cart item couldn't exist without product, the cart item has one to one (or many-to-one if will decide to update) relationship with product so you are not allowed by RDBMS to create a cart item record where point to unexisted product record and you will not be allowed by RDBMS to remove a product record if at least one cart item record which point to given product record exists.

Reply:
fixed ✅